pkg/controller/clusterpool/clusterpool_controller.go

```diff
@@ -973,6 +977,12 @@ func (r *ReconcileClusterPool) reconcileDeletedPool(pool *hivev1.ClusterPool, lo
 			return errors.Wrap(err, "could not delete ClusterDeployment")
 		}
 	}
+	cdcs, err := getAllCustomizationsForPool(r.Client, pool, logger)
+	if err != nil {
+		return err
+	}
+	// If CDC is shared by other clusterpool then the deletion should trigger other clusterpools to resolve any return the finalizer
```

Comment:
This sentence isn't parsing for me. Did you mean:
```suggestion
// If CDC is shared by other clusterpools then the deletion should trigger other clusterpools to reconcile and restore the finalizer
```
?

However, you're only doing this to the local copy as currently written. You would need to `Update()` each CDC to make the change stick.

But I don't think we want to delete the finalizer from the actual object in this way.

I think I understand that we're trying to prevent CDCs from getting stuck deleting. The problem is this scenario:
- CDC listed in two pools' inventories.
- User deletes the common CDC. It is held on the finalizer.
- User deletes one of the pools.
- This code runs (assuming you add the `Update()` and removes the finalizer from the common CDC.
- k8s purges the CDC, since it's got a deletion timestamp but no finalizers.
- The other pool reconciles -- but the CDC is already gone, so it can't restore the finalizer. Instead the CDC just comes up Missing.

Once again you've unearthed an issue that we didn't think of that needs to be solved.

The solution that comes to mind is to make each pool's CDC finalizer unique (e.g. named after the pool) -- so the common CDC in the above scenario would actually start off with _two_ finalizers, and this code would only flush out the one belonging to the pool being deleted. The other finalizer would stick, preventing the deletion.

WDYT?

Reply:
The original bug was that you couldn't delete the CDC once it was registered by a clusterpool. While fixing that bug I was thinking about the shared CDC scenario.

> The solution that comes to mind is to make each pool's CDC finalizer unique (e.g. named after the pool) -- so the common CDC in the above scenario would actually start off with two finalizers, and this code would only flush out the one belonging to the pool being deleted. The other finalizer would stick, preventing the deletion.

I was reluctant of this solution because it adds complexity and the danger of unmanaged finalizers, but it is the obvious solution and there is a problem that I didn't solve.